pyerrors/fits.py

```diff
@@ -656,6 +690,189 @@ def chisqfunc_compact(d):
     return output
 
 
+def _combined_fit(x, y, func, silent=False, **kwargs):
+
+    if kwargs.get('correlated_fit') is True:
+        raise Exception("Correlated fit has not been implemented yet")
+
+    output = Fit_result()
+    output.fit_function = func
+
+    if kwargs.get('num_grad') is True:
+        jacobian = num_jacobian
+        hessian = num_hessian
+    else:
+        jacobian = auto_jacobian
+        hessian = auto_hessian
+
+    x_all = np.concatenate([np.array(o) for o in x.values()])
+    y_all = np.concatenate([np.array(o) for o in y.values()])
+
+    y_f = [o.value for o in y_all]
+    dy_f = [o.dvalue for o in y_all]
+
+    if len(x_all.shape) > 2:
+        raise Exception('Unknown format for x values')
+
+    # number of fit parameters
+    n_parms_ls = []
+    for key in func.keys():
+        if not callable(func[key]):
+            raise TypeError('func (key=' + key + ') is not a function.')
+        if len(x[key]) != len(y[key]):
+            raise Exception('x and y input (key=' + key + ') do not have the same length')
+        for i in range(42):
+            try:
+                func[key](np.arange(i), x_all.T[0])
+            except TypeError:
+                continue
+            except IndexError:
+                continue
+            else:
+                break
+        else:
+            raise RuntimeError("Fit function (key=" + key + ") is not valid.")
+        n_parms = i
+        n_parms_ls.append(n_parms)
+    n_parms = max(n_parms_ls)
+    if not silent:
+        print('Fit with', n_parms, 'parameter' + 's' * (n_parms > 1))
+
+    if 'initial_guess' in kwargs:
+        x0 = kwargs.get('initial_guess')
+        if len(x0) != n_parms:
+            raise Exception('Initial guess does not have the correct length: %d vs. %d' % (len(x0), n_parms))
+    else:
+        x0 = [0.1] * n_parms
+
+    def chisqfunc(p):
+        chisq = 0.0
+        for key in func.keys():
+            x_array = np.asarray(x[key])
+            model = anp.array(func[key](p, x_array))
+            y_obs = y[key]
+            y_f = [o.value for o in y_obs]
+            dy_f = [o.dvalue for o in y_obs]
+            C_inv = np.diag(np.diag(np.ones((len(x_array), len(x_array))))) / dy_f / dy_f
+            chisq += anp.sum((y_f - model) @ C_inv @ (y_f - model))
```

Comment:
An alternative solution would be to create a list of functions a la 
``` python
func_list = np.concatenate([[func[k]] * len(x[k]) for k in keylist])
```
and to define the model via
``` python
model = anp.array([func_list[i](p, x_all[i]) for i in range(len(x_all))])
```
and then combine it with `y` as it is done in the standard routine. 
In this way, you would not have to do create all the arrays and lists over and over again. However, you don't use the feature to call the function with an array of values, which costs probably some time.

As third solution, you could ensure that the values of `x` are arrays at the beginning of the function and create dictionaries for `y_f` and `dy_f` that match `y` such that you can refer to them in this routine to ensure maximum performance.